src/BaseRewardStreams.sol

```diff
@@ -0,0 +1,626 @@
+    }
+
+    /// @notice Returns the end timestamp for a given epoch.
+    /// @dev The end timestamp is just after its epoch, and is the same as the start timestamp from the next epoch.
+    /// @param epoch The epoch to get the end timestamp for.
+    /// @return The end timestamp for the given epoch.
+    function getEpochEndTimestamp(uint48 epoch) public view override returns (uint48) {
+        return uint48(getEpochStartTimestamp(epoch) + EPOCH_DURATION);
+    }
+
+    /// @notice Increases the reward token amounts for a specific rewarded token.
+    /// @param rewarded The address of the rewarded token.
+    /// @param reward The address of the reward token.
+    /// @param startEpoch The starting epoch to increase the reward token amount for.
+    /// @param amounts The token amounts to increase by.
+    function increaseRewardAmounts(
+        address rewarded,
+        address reward,
+        uint48 startEpoch,
+        uint128[] memory amounts
+    ) internal virtual {
+        mapping(uint256 => uint128[EPOCHS_PER_SLOT]) storage storageAmounts = distributions[rewarded][reward].amounts;
+
+        for (uint48 i = 0; i < amounts.length; ++i) {
+            // Overflow safe because `totalRegistered <= type(uint144).max / SCALER < type(uint128).max`.
+            unchecked {
+                uint48 epoch = startEpoch + i;
+                storageAmounts[epoch / EPOCHS_PER_SLOT][epoch % EPOCHS_PER_SLOT] += amounts[i];
+            }
+        }
+    }
+
+    /// @notice Claims the earned reward for a specific account, rewarded token, and reward token, and transfers it to
+    /// the recipient.
+    /// @dev If recipient is address(0) or there is no reward to claim, this function does nothing.
+    /// @param account The address of the account claiming the reward.
+    /// @param rewarded The address of the rewarded token.
+    /// @param reward The address of the reward token.
+    /// @param recipient The address to which the claimed reward will be transferred.
+    function claim(address account, address rewarded, address reward, address recipient) internal virtual {
+        if (recipient == address(0)) revert InvalidRecipient();
+
+        EarnStorage storage accountEarned = accounts[account][rewarded].earned[reward];
+        uint128 amount = accountEarned.claimable;
+
+        // If there is a reward token to claim, transfer it to the recipient and emit an event.
+        if (amount != 0) {
+            DistributionStorage storage distributionStorage = distributions[rewarded][reward];
+            uint128 totalRegistered = distributionStorage.totalRegistered;
+            uint128 totalClaimed = distributionStorage.totalClaimed;
+            uint128 newTotalClaimed = totalClaimed + amount;
+
+            assert(totalRegistered >= newTotalClaimed);
+
+            distributionStorage.totalClaimed = newTotalClaimed;
+            accountEarned.claimable = 0;
+
+            IERC20(reward).safeTransfer(recipient, amount);
```

Comment:
Unless it's difficult or adds a lot of dependencies, I think it may be a good idea. It might be natural to try to withdraw to the same address you are using to earn the rewards.